cin_validator/rules/cin2022_23/rule_8841.py

```diff
@@ -0,0 +1,279 @@
+from typing import Mapping
+
+import pandas as pd
+
+from cin_validator.rule_engine import CINTable, RuleContext, rule_definition
+from cin_validator.test_engine import run_rule
+from cin_validator.utils import make_census_period
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+
+ChildProtectionPlans = CINTable.ChildProtectionPlans
+CPPstartDate = ChildProtectionPlans.CPPstartDate
+LAchildID = ChildProtectionPlans.LAchildID
+CPPID_CPP = ChildProtectionPlans.CPPID
+
+Reviews = CINTable.Reviews
+CPPreviewDate = Reviews.CPPreviewDate
+CPPID_reviews = Reviews.CPPID
+
+# define characteristics of rule
+@rule_definition(
+    # write the rule code here
+    code=8841,
+    # replace ChildProtectionPlans with the value in the module column of the excel sheet corresponding to this rule .
+    # Note that even if multiple tables are involved, one table will be named in the module column.
+    module=CINTable.ChildProtectionPlans,
+    # replace the message with the corresponding value for this rule, gotten from the excel sheet.
+    message="The review date cannot be on the same day or before the Child protection Plan start date.",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[
+        CPPstartDate,
+        CPPreviewDate,
+    ],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # PREPARING DATA
+
+    # Replace ChildProtectionPlans with the name of the table you need.
+    df_cpp = data_container[ChildProtectionPlans].copy()
+    df_reviews = data_container[Reviews].copy()
+
+    # Before you begin, rename the index so that the initial row positions can be kept intact.
+    df_cpp.index.name = "ROW_ID"
+    df_reviews.index.name = "ROW_ID"
+
+    # Resetting the index causes the ROW_IDs to become columns of their respective DataFrames
+    # so that they can come along when the merge is done.
+    df_cpp.reset_index(inplace=True)
+    df_reviews.reset_index(inplace=True)
+
+    # lOGIC
+    # Implement rule logic as described by the Github issue.
+    # Put the description as a comment above the implementation as shown.
+
+    # Within a <ChildProtectionPlans> group, there should be no <CPPreviewDate> (N00116) that is the same as or before the <CPPstartDate> (N00105)
+    # Issues dfs should return rows where CPPreviewDate is less than or equal to the CPPstartDate
+
+    #  Create dataframes which only have rows with CP plans, and which should have one plan per row.
+    df_cpp = df_cpp[df_cpp[CPPstartDate].notna()]
+    df_reviews = df_reviews[df_reviews[CPPreviewDate].notna()]
+
+    #  Merge tables to get corresponding CP plan group and reviews
+    df_merged = df_cpp.merge(
+        df_reviews,
+        left_on=["CPPID", "LAchildID"],
+        right_on=["CPPID", "LAchildID"],
```

Comment:
Almost there.

If you check the sample dataframes file (https://github.com/data-to-insight/CIN-validator/blob/main/cin_validator/demo_ingress_class.ipynb) or see the table column lists in api.py, you'll notice that in addition to "CPPID" and  "LAchildID", these tables also have to be grouped by CINdetailsID. 

So that has to be included in your join too if not two CPP groups that belong to different CINdetails groups but have the same ID would be seen as if they're the same.